bpf: switch to cilium/ebpf loader and native netlink calls

ti-mo opened the pull request:
Description taken from the main commit. Best reviewed per commit.

> This commit mainly modifies replaceDatapath() to invoke the cilium/ebpf ELF
loader instead of shelling out to iproute2's tc and ip commands.
>
> Note that this patch focuses on a 1:1 conversion to minimize churn and to keep
it reasonably reviewable. Many changes of larger magnitude are planned to make
the datapath and loader provide an actual API that will serve as a foundation
for testing infrastructure and higher-level packages. The goal here is to get
the ball rolling.
>
> Programs are no longer being referred to using their ELF section names, but
using their C function names instead.
>
> Changes to init.sh are notably excluded because they require additional care
and attention, as well as adding CLI commands to the agent that wrap the Go
loader code to replace tc/ip invocations. This will be tackled in follow-ups.
>
> ctx and `xdp bool` were removed from replaceDatapath(). ctx is no longer used,
ebpf library code is generally not cancellable. All uses of xdp/xdpMode were
either 'true, xdpMode' or 'false, ""', so this was reduced to a single flag
xdpMode that is acted upon when non-zero. This simplifies logic and doesn't
allow both parameters to contradict each other.

```release-note
Load datapath ELFs using cilium/ebpf
```

ti-mo committed:
collection: wrap ebpf.Collection(Spec) in iproute2 compat layer

As documented in the code, this patch adds wrappers around LoadCollectionSpec()
and LoadCollection() to package bpf.

For compatibility with iproute2 bpf_elf_map definitions, iproute2Compat()
parses the unsupported extra tail bytes and interprets them to set the pinning
flag and to attribute 'X/Y' sections containing tail calls to their respective
map IDs and the indices within them.

classifyProgramTypes() automatically detects attach types of Cilium's ELFs
based on the names of the programs it contains.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: inline global variables in Go userspace

inlineGlobalData() was added to inline map loads from .data into the bytecode
in userspace. Once all references have been inlined, the map is removed from
the spec to skip loading it into the kernel.

Later on, global constants can be declared const so they end up in .rodata,
where they can be found by CollectionSpec.RewriteConstants, which will replace
the ELF templating logic.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf,datapath: switch to cilium/ebpf loader and native netlink calls

This commit mainly modifies replaceDatapath() to invoke the cilium/ebpf ELF
loader instead of shelling out to iproute2's tc and ip commands.

Note that this patch focuses on a 1:1 conversion to minimize churn and to keep
it reasonably reviewable. Many changes of larger magnitude are planned to make
the datapath and loader provide an actual API that will serve as a foundation
for testing infrastructure and higher-level packages. The goal here is to get
the ball rolling.

Programs are no longer being referred to using their ELF section names, but
using their C function names instead.

Changes to init.sh are notably excluded because they require additional care
and attention, as well as adding CLI commands to the agent that wrap the Go
loader code to replace tc/ip invocations. This will be tackled in follow-ups.

`xdp bool` was removed from replaceDatapath(). All uses of xdp/xdpMode were
either 'true, xdpMode' or 'false, ""', so this was reduced to a single flag
xdpMode that is acted upon when non-zero. This simplifies logic and doesn't
allow both parameters to contradict each other.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: remove parseExtra() from bpffs migrations

Start/FinalizeBPFFSMigration now take a *ebpf.CollectionSpec, which should be
obtained using bpf.LoadCollectionSpec. This is a wrapper that contains iproute2
compatibility logic, including draining MapSpec.Extra, so this is no longer
necessary.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
datapath: allow distinguishing JoinEP warnings

Previously, different call sites emitted the same warning message,
making it difficult to trace back where the error originated from.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
init,cleanup: remove TC filters containing 'cilium' in their names

With the addition of Go code that loads and attaches BPF programs,
we're no longer using the BPF file/section as the tc filter name.

Assume the filter names can also contain 'cilium'.

Signed-off-by: Timo Beckers <timo@isovalent.com>

michi-covalent reviewed (approved):
reviewed pkg/option/config.go 👍

joestringer commented on pkg/bpf/collection.go:
```diff
@@ -0,0 +1,251 @@
+	// By default, allocate a 1MiB verifier log buffer if first load attempt
+	// fails. This was adjusted around Cilium 1.11 for fitting bpf_lxc insn
+	// limit messages.
+	if opts.Programs.LogSize == 0 {
```
I seem to recall that if you provide too small of a buffer, then the kernel may reject the program during verification purely because it ran out of verifier log buffer, even if it would otherwise successfully verify the program without the buffer. For this reason I think the desired behaviour would be first to attempt to load with no buffer, then only if that fails, load with a buffer, then also potentially check whether the buffer was big enough to gather proper verifier error output, and possibly reload multiple times with bigger and bigger buffers until we can guarantee that we gather the full verifier log.

joestringer replied:
^ IIRC iproute2 includes all of the above logic, so to import the equivalent behaviour here we'll need to write the loop & retry logic ourselves.

ti-mo replied:
The library already incorporates this functionality, see https://github.com/cilium/ebpf/blob/629483784baa95e91c992134241439525cad0dac/prog.go#L274-L304.

By default, prog_load is run with log level 0, so there will be no verifier output. If that fails, it's retried with the verifier log enabled, and the buffer set to the default size or the LogSize specified in opts.

If the log level is non-zero and a buffer is provided, log full means -ENOSPC, but that's only when the program fails verification.

The problem with retrying with bigger and bigger buffers is that we can't really estimate how big the log will be. Given a default starting buffer of 65k, where do we go from there? Do we double it? 1MiB is a few retries off from there.. How often do we retry and when do we give up? I don't believe retrying with bigger logs is a good default strategy, at least not with bpf_lxc spending ~600ms in the verifier on my machine.

Instead, the user should provide a reasonable size. As mentioned in the comment, largest log I've seen by far is from bpf_lxc and was a bit under a MiB. Verifier errors should be comparatively rare outside of development, so if ENOSPC is encountered, it would most likely be caught before going to production. At that point, this value can be bumped. Rinse, repeat.

joestringer replied:
That makes total sense, thanks for the explanation.

joestringer replied:
btw minor nit here, the code specifies 1<<24 (16MiB) rather than 1<<20 (1MiB).

ti-mo replied:
Good catch. 1<<24 is 1000000 in hex, not decimal. :sweat:

joestringer commented on pkg/bpf/collection.go:
```diff
@@ -0,0 +1,251 @@
+	}
+
+	if dl := len(data.Contents); dl != 1 {
+		return nil
```
Should this perhaps be something like `return fmt.Errorf("expected one key in .rodata, found %d", n)`?

(Possibly also other related error condition in cast below, matching the upstream library's `collection.go`)

joestringer commented on pkg/bpf/collection.go:
```diff
@@ -0,0 +1,251 @@
+		}
+	}
+
+	return nil
```
As a sanity check, we could also consider tracking & warning when there is info in `.data` but it's unused by any of the programs. Might help point out some bugs for developers. Nice to have, not required though.

ti-mo replied:
If there are unused vars, the compiler should warn/error. If there are globals marked `unused`, that's usually for a reason, so wouldn't this kind of warning just add noise?

I've found with this kind of code it's better to either error hard, or automatically do the right thing instead of throwing a warning. There's so much output these tend to get lost.

joestringer replied:
:+1: Fair enough yeah. Agree around automatically doing the right thing. My initial thought was that I couldn't think of any reasons for `If there are globals marked unused, that's usually for a reason`, but now I think more it seems like it could be a convenient way to just stuff some data in an ELF if you need it. So no need for any follow up here.

joestringer commented on pkg/datapath/loader/netlink.go:
```diff
@@ -69,85 +66,153 @@ func replaceQdisc(ifName string) error {
+
+		return nil, fmt.Errorf("program %s: %w", progName, err)
 	}
 
```
nit: We should probably avoid shadowing `err` from the `attachProgram` ~here~ in the `Failed to revert bpffs map migration` error return path (line 112/118). We could just declare the finalize err as a second error, log that one if it fails, then continue to line 121 to return the actual load error. I realize the code already does this shadowing, but while we're looking at this it'd be nice to improve it.

joestringer replied:
^ I noticed this in `graftDatapath()` as well, there's probably room to refactor and merge some more of those functions but we can always do that later.

joestringer commented on pkg/datapath/loader/netlink.go:
```diff
@@ -69,85 +66,153 @@ func replaceQdisc(ifName string) error {
+
+	if err := netlink.FilterReplace(filter); err != nil {
+		return fmt.Errorf("replacing tc filter: %w", err)
+	}
```
:tada: I've been looking forward to named programs. I wonder whether we should come up with a different naming scheme rather than using the device name (since it's not entirely trivial to backtrack from device name to the endpointID and rest of the info on the system), but we can argue about that in subsequent PRs, these are minor nits.

ti-mo replied:
Yes! Related convo's: https://github.com/cilium/cilium/pull/19735#discussion_r888924559 and https://github.com/cilium/cilium/issues/10067#issuecomment-1024322477.

joestringer commented on pkg/datapath/loader/netlink.go:
```diff
@@ -69,85 +66,153 @@ func replaceQdisc(ifName string) error {
+		Name:         fmt.Sprintf("cilium-%s", link.Attrs().Name),
+		DirectAction: true,
+	}
+
```
One of the previous commits changed TCFilterPriority to a uint32, and yet this code is now truncating it to 16 bits. Seems like maybe we should be limiting it to uint16 instead?

ti-mo replied:
This was changed to a u32 to prevent the user from being able to specify negative values and because there's no u16 equivalent of `viper.GetUint32()`.

What's your understanding of truncating vs. limiting? I would suggest input validation against `math.MaxUint16` that makes the agent bail out early. Before this PR, an error would've occurred when trying to call tc during prog attach, while my current proposal truncates silently, which might be surprising.

joestringer replied:
I see. Yes, I like your idea around input validation early in the option package using `math.MaxUint16` and we can even fatal out if the user provides invalid input like that. In this sort of case what I would prefer is to get the value into the correct type as soon as possible so that only the option package needs to worry about this viper "no GetUint16() method" issue. Then everywhere else in the daemon can correctly assume the correct type rather than having casts like this scattered around.

joestringer commented on pkg/datapath/loader/netlink.go:
```diff
@@ -69,85 +66,153 @@ func replaceQdisc(ifName string) error {
+		},
+		Fd:           prog.FD(),
+		Name:         fmt.Sprintf("cilium-%s", link.Attrs().Name),
+		DirectAction: true,
```
Is there some magic to the `netlink.MakeHandle(0, 1)` that we should be aware of here? I couldn't trivially find the corresponding encoding in iproute2 to confirm that this does the same thing.

ti-mo replied:
Fair point, couldn't find anything significant either, this was copypasted from `ipvlan.go`. At least `tc` allows using the full 32 bits of `Handle`, so I guess we don't need `MakeHandle` here.

```
# tc filter replace dev cilium-probe egress prio 1 handle 0xffffffff bpf da obj bpf_network.o sec from-network
# tc filter show dev cilium-probe egress
filter protocol all pref 1 bpf chain 0 handle 0xffffffff bpf_network.o:[from-network] direct-action not_in_hw id 2461 tag dbc39707948386e6 jited
```

joestringer commented on vendor/github.com/prometheus/client_golang/prometheus/collectors/collectors.go:
```diff
@@ -14,3 +14,27 @@
 // conveniently collect process and Go-related metrics.
 package collectors
+
+import "github.com/prometheus/client_golang/prometheus"
```
This commit seems unrelated to the rest of the PR. Was it included by mistake?

ti-mo replied:
Unrelated yes, but this addresses a regression that popped up again after a master rebase. (https://github.com/cilium/cilium/issues/19425) I just pulled it in here to be able to put this through CI. client_golang v1.12.1 causes the agent's API handler to panic on trying to encode NaNs as json. I'm not sure why this is working fine on master. We should definitely upgrade client_golang to v1.12.2 before cutting Cilium 1.12, since 1.12.1 greatly inflates cardinality.

This obviously needs to go in as a separate PR. :) 

Edit: I see this was recently done in 044681a8540ba5cfc39b400bfc79a3c1723f9f13. Still needs a follow-up to re-add it to dependabot: https://github.com/cilium/cilium/pull/20075.

ti-mo replied:
Commit removed.

joestringer reviewed (changes requested):
Changes look good overall, except maybe stuff like the `TCFilterPriority` type change which introduces type casts u32->u16. I also don't see how the prometheus change is related to the PR. Also, I think that today in adverse cases, we will retry loading programs until we get a valid verifier log, which is super important for debugging, otherwise we have no guarantee that 1MB is enough buffer to load the program or grab the verification errors, and we may end up with a failing load and no hints why.

Beyond that, I notice these errors in the 4.9 test run:

```
Top 5 errors/warnings:
Error while rewriting endpoint BPF program
endpoint regeneration failed
Creation of endpoint failed
Cancelling obsolete endpoint creating due to new create for same pod
JoinEP: Failed to attach ingress program
```

This may actually just be a new representation of #16928 .. full logs:

```
2022-05-31T09:02:31.425134491Z level=warning msg="Cancelling obsolete endpoint creating due to new create for same pod" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=0 ipv4= ipv6= k8sPodName=/ subsys=daemon
2022-05-31T09:02:33.160278753Z level=warning msg="JoinEP: Failed to attach ingress program" containerID= datapathPolicyRevision=0 desiredPolicyRevision=11 endpointID=2008 error="program handle_xgress: replacing clsact qdisc for interface lxcefcbc3fe423f: no such device" file-path=2008_next/bpf_lxc.o identity=2424 ipv4= ipv6= k8sPodName=/ subsys=datapath-loader veth=lxcefcbc3fe423f
2022-05-31T09:02:33.160312779Z level=error msg="Error while rewriting endpoint BPF program" containerID= datapathPolicyRevision=0 desiredPolicyRevision=11 endpointID=2008 error="program handle_xgress: replacing clsact qdisc for interface lxcefcbc3fe423f: no such device" identity=2424 ipv4= ipv6= k8sPodName=/ subsys=endpoint
2022-05-31T09:02:33.160605118Z level=warning msg="generating BPF for endpoint failed, keeping stale directory." containerID= datapathPolicyRevision=0 desiredPolicyRevision=11 endpointID=2008 file-path=2008_next_fail identity=2424 ipv4= ipv6= k8sPodName=/ subsys=endpoint
2022-05-31T09:02:33.160621598Z level=warning msg="Regeneration of endpoint failed" bpfCompilation=0s bpfLoadProg=1m33.168181964s bpfWaitForELF="2.275µs" bpfWriteELF="370.065µs" containerID= datapathPolicyRevision=0 desiredPolicyRevision=11 endpointID=2008 error="program handle_xgress: replacing clsact qdisc for interface lxcefcbc3fe423f: no such device" identity=2424 ipv4= ipv6= k8sPodName=/ mapSync="56.625µs" policyCalculation="40.616µs" prepareBuild="279.359µs" proxyConfiguration="4.911µs" proxyPolicyCalculation="39.764µs" proxyWaitForAck=0s reason="updated security labels" subsys=endpoint total=1m33.171392964s waitingForCTClean="821.046µs" waitingForLock="135.542µs"
2022-05-31T09:02:33.162713435Z level=error msg="endpoint regeneration failed" containerID= datapathPolicyRevision=0 desiredPolicyRevision=11 endpointID=2008 error="program handle_xgress: replacing clsact qdisc for interface lxcefcbc3fe423f: no such device" identity=2424 ipv4= ipv6= k8sPodName=/ subsys=endpoint
2022-05-31T09:02:33.162796213Z level=warning msg="Creation of endpoint failed" containerID= datapathPolicyRevision=0 desiredPolicyRevision=11 endpointID=2008 error="timeout while waiting for initial endpoint generation to complete: context canceled" identity=2424 ipv4= ipv6= k8sPodName=/ subsys=daemon
```

The 4.19 job is similar:

```
2022-05-31T08:54:53.442036066Z level=warning msg="Cancelling obsolete endpoint creating due to new create for same pod" containerID= datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=0 ipv4= ipv6= k8sPodName=/ subsys=daemon
2022-05-31T08:55:01.630906385Z level=warning msg="JoinEP: Failed to attach ingress program" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=629 error="program handle_xgress: replacing clsact qdisc for interface lxcbf85225b1b37: no such device" file-path=629_next/bpf_lxc.o identity=46251 ipv4= ipv6= k8sPodName=/ subsys=datapath-loader veth=lxcbf85225b1b37
2022-05-31T08:55:01.630929077Z level=error msg="Error while rewriting endpoint BPF program" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=629 error="program handle_xgress: replacing clsact qdisc for interface lxcbf85225b1b37: no such device" identity=46251 ipv4= ipv6= k8sPodName=/ subsys=endpoint
2022-05-31T08:55:01.630995341Z level=warning msg="generating BPF for endpoint failed, keeping stale directory." containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=629 file-path=629_next_fail identity=46251 ipv4= ipv6= k8sPodName=/ subsys=endpoint
2022-05-31T08:55:01.631048430Z level=warning msg="Regeneration of endpoint failed" bpfCompilation=0s bpfLoadProg=1m49.851509824s bpfWaitForELF="4.061µs" bpfWriteELF=1.00146ms containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=629 error="program handle_xgress: replacing clsact qdisc for interface lxcbf85225b1b37: no such device" identity=46251 ipv4= ipv6= k8sPodName=/ mapSync="73.697µs" policyCalculation="42.472µs" prepareBuild="212.464µs" proxyConfiguration="3.864µs" proxyPolicyCalculation="38.907µs" proxyWaitForAck=0s reason="updated security labels" subsys=endpoint total=1m49.856346093s waitingForCTClean=1.392366ms waitingForLock=455ns
2022-05-31T08:55:01.631155721Z level=error msg="endpoint regeneration failed" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=629 error="program handle_xgress: replacing clsact qdisc for interface lxcbf85225b1b37: no such device" identity=46251 ipv4= ipv6= k8sPodName=/ subsys=endpoint
2022-05-31T08:55:01.634594785Z level=warning msg="Creation of endpoint failed" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=629 error="timeout while waiting for initial endpoint generation to complete: context canceled" identity=46251 ipv4= ipv6= k8sPodName=/ subsys=daemon
```

markpash commented on cilium/cmd/cleanup.go:
```diff
@@ -486,11 +486,14 @@ func getTCFilters(link netlink.Link) ([]*netlink.BpfFilter, error) {
 			if bpfFilter, ok := f.(*netlink.BpfFilter); ok {
+				// Filters created Go bpf loader are of format 'cilium-<iface>'.
+				// iproute2 would use the filename and section, e.g. bpf_overlay.o:[from-overlay].
 				if strings.Contains(bpfFilter.Name, "bpf_netdev") ||
```
Nit: perhaps a regex match is simpler?
Naive attempt: `^bpf_(netdev|network|host|lxc|overlay)|cilium.*`
And then have a global `regex.Regexp` created with `regexp.MustCompile()`

ti-mo replied:
Good suggestion, but from a maintainability perspective, it really helps if those `bpf_*` entries to turn up in greps.

joestringer commented on pkg/option/config.go:
```diff
@@ -2231,7 +2232,7 @@ type DaemonConfig struct {
 	// TCFilterPriority sets the priority of the cilium tc filter, enabling other
 	// filters to be inserted prior to the cilium filter.
-	TCFilterPriority int
+	TCFilterPriority uint32
```
Following up on our previous discussion, can we switch this to uint16 and just do the validation in the `Populate()` function?

joestringer commented on pkg/datapath/loader/compile.go:
```diff
@@ -363,8 +363,8 @@ func compileDatapath(ctx context.Context, dirs *directoryInfo, isHost bool, logg
 					WithError(err).Debug("JoinEP: Failed to compile")
+				return err
 			}
-			return err
```
:thinking: Can we follow up on this specific change in the original PR (https://github.com/cilium/cilium/pull/18267) just to separate the refactor vs. the fix? This change by itself seems very nuanced, in addition to the rest of this PR's potential impact.

In particular, I see that I had commented on this line in the original PR and now that I look at it, it seems like this just ignores compile errors as long as the error occurs within the context timeout.

joestringer replied:
^^ I just noticed that the review was draft (because that PR itself is still draft), so I posted the comment just now.

ti-mo replied:
I've dropped the changes to `compileDatapath()` from the commit. Running another round of tests.

ti-mo replied:
@joestringer The good news is context cancellation now propagates to the loader code as intended, so we should be able to bail out properly now. However, the changes to `compileDatapath()` do cause trouble down the line:

```
cilium-combined-logs.txt:2022-07-06T14:28:19.381956708Z level=error msg="Command execution failed" cmd="[llc -mattr=dwarfris -march=bpf -mcpu=v2 -filetype=obj -o /var/run/cilium/state/templates/f2716579d3201fb10b2c21d3cd45087ab44160473fa1f3178ae515214c30fdca/bpf_lxc.o]" error="signal: killed" subsys=datapath-loader
cilium-combined-logs.txt:2022-07-06T14:28:19.381958982Z level=error msg="Failed to compile bpf_lxc.o: Command execution failed for [llc -mattr=dwarfris -march=bpf -mcpu=v2 -filetype=obj -o /var/run/cilium/state/templates/f2716579d3201fb10b2c21d3cd45087ab44160473fa1f3178ae515214c30fdca/bpf_lxc.o]: context canceled" compiler-pid=621 linker-pid=622 subsys=datapath-loader
cilium-combined-logs.txt:2022-07-06T14:28:19.381991491Z level=error msg="Error while rewriting endpoint BPF program" containerID= datapathPolicyRevision=0 desiredPolicyRevision=1 endpointID=3847 error="failed to parse ELF file /var/run/cilium/state/templates/f2716579d3201fb10b2c21d3cd45087ab44160473fa1f3178ae515214c30fdca/bpf_lxc.o: unable to open ELF: EOF" identity=4 ipv4= ipv6= k8sPodName=/ subsys=endpoint
```

The last line `failed to parse ELF file ...` is repeated hundreds of times because the pipeline now thinks compilation succeeded when it was actually cancelled. (this happens rather often in CI VMs under load)

Attached one such failure here: [0632a6d8_K8sKafkaPolicyTest_Kafka_Policy_Tests_KafkaPolicies.zip](https://github.com/cilium/cilium/files/9064184/0632a6d8_K8sKafkaPolicyTest_Kafka_Policy_Tests_KafkaPolicies.zip)

Failures during Kafka deploy/image pull seem _really_ common, this leads me to think they might be load-induced.

aditighag replied:
@ti-mo There was a reason I made those changes around returning errors  - https://github.com/cilium/cilium/pull/18267#issuecomment-996254405. The agent code path at many places isn't checking for cancelled context, and that's why they log the error messages even when the failures are expected . :(

joestringer reviewed (changes requested):
I took another brief look, just a couple of remaining comments. Overall it looks good.

ti-mo committed:
endpoint: execute aliveCancel() before draining event queue

The context passed to ParseExternalRegenerationMetadata, which is the
one that ends up propagating all the way down to ReloadDatapath and
friends, can be Endpoint.aliveCtx if the ExternalRegenerationMetadata
does not contain a parent context.

Make sure this context is cancelled early on in the Endpoint teardown
process to allow downstream logic (bpf prog load, netlink attach, ..)
to bail out instead of trying to attach to resources that are already
removed by the container runtime.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
endpoint: gate regeneration errors behind context cancellation checks

When the CNI plugin gives up on creating an Endpoint, or more generally,
when a Context passed into a loader function gets canceled, many functions
in the call chain will return error.

To avoid logging these errors as they occur in each part of the call chain,
only emit them when they don't contain a context.Canceled. Exceeded deadlines
should still be logged.

Aside from cleaning up the logs, this also causes the CI log scanner to not
trip over so many potential `level=error` lines that occur due to (CPU)
resource constraints when compiling/loading/attaching many endpoint progs.

Signed-off-by: Timo Beckers <timo@isovalent.com>

joestringer commented on pkg/endpoint/restore.go:
```diff
@@ -76,7 +76,7 @@ func ReadEPsFromDirNames(ctx context.Context, owner regeneration.Owner, policyGe
 			})
 			fullDirName := filepath.Join(basePath, epDirName)
-			scopedLog.Warning(fmt.Sprintf("Found incomplete restore directory %s. Removing it...", fullDirName))
+			scopedLog.Info(fmt.Sprintf("Found incomplete restore directory %s. Removing it...", fullDirName))
```
I think this level change is probably reasonable considering that the endpoint was in the process of being regenerated and the directory swap was not completed when the prior agent was shut down; for the current agent run, there's not a whole lot to do about this situation.

For context I'd say the reason this is a warning today is that if a previous cilium-agent run got into a state where regenerating an endpoint was continuously unsuccessful, then it's plausible that that endpoint with the problem is not being correctly provisioned, which indicates a bug. Then for this run, we would try to raise the signal that perhaps this endpoint is not being correctly handled during this run. Furthermore, if this somehow is the last state that Cilium has for this endpoint then this could be the last chance for Cilium to signal to the user that it's forgetting about the endpoint and probably some action needs to be taken.

That said, maybe we can do a better job of solving the reliability problem here at a higher layer by introducing a periodic checker that validates each k8s Pod scheduled to this node has a corresponding endpoint. Then that checker could either attempt to fix any issue or do the signalling instead.

I think the one key thing that I'm wondering about here though is that I would expect that "info" logs should be emitted only for "normal" events. Incomplete restore sounds like the prior cilium-agent run was prematurely terminated and it didn't clean up all its state correctly. This is not a normal operation. From that perspective, I'd be inclined to leave this at Warning level.

joestringer replied:
If the motivation here is to avoid triggering the CI failures due to cleaning up this directory, then I can understand where it's coming from. If we're hitting this frequently in today's CI then maybe we can temporarily add this to the list of logs to exclude from failing errors/warnings, then schedule a followup to investigate (cc @borkmann). On the other hand if this is not typically seen on the master branch and suddenly this PR makes it come up frequently, then there may be something this PR does to make BPF prog loads to fail during the prior run and this is the only signal we're getting for that.

ti-mo replied:
The biggest behavioural shift in this PR is executing `Endpoint.aliveCancel()` early on in the `Endpoint.Stop()` process. Before, it seems like Cilium would happily go on to compile, rewrite and (try to) attach endpoint programs even though the CNI already gave up and the veths were long gone. This is more optimal, but could also surface a few new behaviours, and perhaps this incomplete dir is related. I assume dirs are left behind if endpoint compilation is cancelled, which sounds pretty normal to me. And after all, the agent cannot prevent itself from being terminated, so whatever on-disk state there is, it needs to tolerate.

I haven't put much thought into this one in particular, but isn't an incomplete dir just one with a `_next` or `_next_fail` suffix? The log message simply says "hey, I'm cleaning up". Also relatively normal. If the agent couldn't continue for whatever reason, the appropriate action would be a `Fatal()`, but I don't think that's the case here. Hence, info.

To me, the distinction between info and error lies between something needing operator intervention or not. From an operator POV, `level=error` in production should trigger an alert and cause someone to get paged. If not, what is its meaning?

Philosophical discussion aside, this case in particular sounds pretty expected, and there's little the agent can do. I think the pre-restart agent logs would contain the appropriate entries pertaining to this particular failure, unless it was SIGKILLed, but that sounds rather unlikely here.

joestringer replied:
> I haven't put much thought into this one in particular, but isn't an incomplete dir just one with a `_next` or `_next_fail` suffix?

Yes. Hopefully there's still a regular directory without the suffix there at that time to allow a safe restore. So yeah I guess that's fine to just clean up. Agree that it should handle the case where the agent is just prematurely terminated; there's not much else we could do there.

> To me, the distinction between info and error lies between something needing operator intervention or not. From an operator POV, level=error in production should trigger an alert and cause someone to get paged. If not, what is its meaning?

:+1: but also the above is referring to `level=error` rather than `level=warning`. One thing I was thinking about is that if we can catch any sort of "unusual but recoverable" cases in CI by logging warnings then catching those in CI, then that can give us signal about areas to look closer, possibly even catching bugs before we release them.

But yeah I guess this is fine to just log at Info level, it's at least a signal in case there is anything weird going on here but there's not any strong signal that additional administrator intervention is required. :+1:

joestringer commented on pkg/serializer/func_queue.go:
```diff
@@ -3,100 +3,52 @@
-
-var (
-	// NoRetry always returns false independently of the number of retries.
-	NoRetry = func(int) bool { return false }
```
If `NoRetry` is unused, please consider putting the deletion into a separate commit. Combining it into the same commit as the other logic here makes it harder for reviewers to focus on the impacts of individual changes.

ti-mo replied:
Not sure what you mean. `NoRetry` was the only value that was ever passed as a `waitFunc` to `Enqueue`, implying it is the default. And since it never retries, that prompted me to remove the retry code. It was never unused, just redundant, and it should remain part of the commit that removed `waitFunc`/`WaitFunc`.

joestringer replied:
Whenever I see a commit that is double-barreled like `datapath: remove unused serializer retry logic, return errors from fc.Wait`, this is an indication to me that there are two logically separate changes that do not need to be in the same commit. Putting them together means that the reviewer needs to track which parts of the commit relate to the first logical change and which relate to the second logical change, and try to identify any changes that don't belong to either (which is often a sign of unintentional changes, which could be a bug). It's just a bit more cognitive work for the reviewer.

The `Wait()`/`Stop()` changes didn't seem to be related to the `NoRetry` change, so while reviewing I went through the above thought process and figured it would have been simpler to understand if `NoRetry` was separate. Similarly if it turns out we miss something, it would be easier to understand where the bug is if we can bisect to one patch with fewer changes.

ti-mo replied:
Sure, I'll try to rejig this. Edit: split the functional change from the retry removal.

joestringer commented on pkg/serializer/func_queue.go:
```diff
@@ -3,100 +3,52 @@
 func NewFunctionQueue(queueSize uint) *FunctionQueue {
 	fq := &FunctionQueue{
-		queue:  make(chan queuedFunction, queueSize),
+		queue:  make(chan func() error),
```
If this commit is going to ignore the `queueSize` parameter, then it should also remove the parameter so that callers do not expect some different behaviour compared to what the API provides.

ti-mo replied:
Thanks, good catch!

joestringer commented on pkg/serializer/func_queue.go:
```diff
@@ -3,100 +3,52 @@
-	if err := ctx.Err(); err != nil {
-		return fmt.Errorf("serializer: %w", err)
+		return
+	default:
```
It looks to me like this last commit refactors a bunch of this datastructure, but the core functional change here would be the same as if `Enqueue` just stored the `err` and `Wait(...)` returned that err instead of returning `nil` at the end of the function.

ti-mo replied:
Correct, this is mentioned in the commit message and subject. Just an observation, or changes required?

joestringer replied:
Yeah just an observation. I struggled quite a bit with this patch during the previous round of review because of how it refactored the entire datastructure and the users rather than making the small functional change described above then doing the datastructure rewrite second. There are ways to split up such changes into smaller, more easily reviewable chunks.

joestringer commented on pkg/serializer/func_queue.go:
```diff
@@ -3,100 +3,52 @@
-// it properly.
-func (fq *FunctionQueue) Enqueue(f func() error, waitFunc WaitFunc) {
-	fq.queue <- queuedFunction{f: f, waitFunc: waitFunc}
+	fq.queue <- f
```
If two goroutines (A,B) call `Enqueue()` within quick succession and the function takes a reasonably long time, then A could send the function to the goroutine launched by `NewFunctionQueue()` and B will block forever. The backing goroutine would read the function from A and execute it, close the `fq.stopCh` and return without reading again the function from B.

I don't think this library is necessarily intended to be threadsafe, but in that case we should probably either document that callers should never call `Enqueue()` without ensuring exclusion, or modify this structure such that concurrent calls to Enqueue don't result in blocking the latter goroutine(s) indefinitely.

ti-mo replied:
Good point, that's not correct. Not much point in doing a nonblocking read from `stopCh`.

The send into `queue` and read from `stopCh` should be part of a simple select without a default clause. `queue` is never closed, so putting the send into the select won't cause a panic. If the queue is closed, all pending callers will exit Enqueue through the read of `stopCh`.

joestringer commented on daemon/cmd/state.go:
```diff
@@ -284,6 +284,7 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) (resto
 		log.WithField(logfields.EndpointID, ep.ID).Info("Successfully restored endpoint. Scheduling regeneration")
 		go func(ep *endpoint.Endpoint, epRegenerated chan<- bool) {
 			if err := ep.RegenerateAfterRestore(); err != nil {
+				log.WithField(logfields.EndpointID, ep.ID).WithError(err).Debug("error regenerating during restore")
```
Now that this is debug, and the callee no longer has any such error message, if there is an error during restore there is no corresponding log message. I didn't look _too_ closely but it didn't seem like this error will ever be printed, so this could make it difficult to pinpoint the cause of a problem during restore. Should this be `Warn()` level, perhaps wrapped in the extra `errors.Is(..., context.Cancelled)` check?

ti-mo replied:
`ep.RegenerateAfterRestore()` returns one of `fmt.Errorf("failed while regenerating endpoint")` or `nil`, so errors aren't bubbled up to the caller. This is due to `Endpoint.Regenerate()` returning a `<-chan bool` and is probably something we should change at some point. This overuse of channels makes the code rather fragile and inflexible to changes.

Would you rather have a `Warn()` behind a check on `aliveCtx` for the time being? Note that as of `Endpoint.Regenerate()`, the context passed down to compile/attach can be either `regenMetadata.ParentContext` or `e.aliveCtx`, so only trusting `aliveCtx` is not technically correct. We need the error value to be sure.

joestringer replied:
Hrm. This is more nuanced than I'd like. It seems that all of the `restoreIdentity()` failures are related to `aliveCtx` so we can safely ignore those. Basically the endpoint is just deleted while we are restoring it, so it's a normal operation not worth logging. Then for the `Endpoint.Regenerate()` case, as you say the error is getting a bit hidden. I'm not sure there's much point in logging the error here after all.

It looks like the underlying regeneration error will be logged to the endpoint's log when it fails, which is at least _something_. It won't end up in the main log if the initial restore fails though. I'm not really sure the best way to handle this right now, other than maybe to have `endpoint.Regenerate()` also log the regeneration error since that function actually receives the underlying regeneration error via the EventQueue's `resChan`.

As for this line of code, I'm fine to leave this at Debug now. Thanks for humouring me on   how to improve this area.

joestringer commented on pkg/datapath/loader/cache.go:
```diff
@@ -288,32 +289,39 @@ func (o *objectCache) fetchOrCompile(ctx context.Context, cfg datapath.EndpointC
 	scopedLog := log.WithField(logfields.BPFHeaderfileHash, hash)
 
 	// Serializes attempts to compile this cfg.
+	// TODO(tb): replace with sync.Once.
```
:thinking: interesting idea. As long as one goroutine triggers the compile and the others can have a mechanism to block on that (plus some sort of reasonable timeout so we can get retries from further up the stack), this could be worth exploring. But yeah this should not be necessary for solving the problems at hand in the current PR.

ti-mo replied:
Yes, a future compilation API would take this into account.

One extra requirement that was overlooked in the existing 'queue' system: multiple endpoints can hash to the same template, but each of those endpoints have independent contexts. A compilation should not be cancelled until all endpoints that want the same hash are cancelled, or cancelling the primary one will cause endpoint compilations/creations to fail for all of them.

joestringer replied:
Yeah good point. It gets a little more tricky with that constraint, just because new endpoints could show up at any time and attempt to add their own context into the mix. To implement the above you would need to support runtime addition of contexts into the list. Probably end up having a whole other goroutine just to handle the context merging and when to actually signal the cancel to the goroutine performing the compile (otherwise, any of the cancelled contexts could interrupt the compile mid-way but there could be new endpoint contexts show up that allow the operation to continue).

joestringer commented on pkg/datapath/loader/cache.go:
```diff
@@ -288,32 +289,39 @@ func (o *objectCache) fetchOrCompile(ctx context.Context, cfg datapath.EndpointC
+	if err := fq.Wait(); err != nil {
+		if !errors.Is(err, context.Canceled) {
+			scopedLog.WithError(err).Warning("Error while waiting for BPF template compilation")
+		}
```
This would be checking whether potentially the `ctx` passed by the _other_ goroutine was cancelled. Then log if that was due to some non-cancellation error (including timeout).

But this is a bit odd, if you have multiple goroutines blocking here, they will all print the same log based on the ctx from the goroutine that does the compile. I don't think that's necessary to log, as this is already logged by the goroutine that does the compile, after the `o.build()` above.
```suggestion
	if err := fq.Wait(); err != nil {
```

joestringer commented on pkg/datapath/loader/netlink.go:
```diff
@@ -66,60 +66,123 @@ func replaceQdisc(ifName string) error {
+	l.Debug("Attaching program to interface")
+	if err := attachProgram(link, coll.Programs[progName], directionToParent(direction), xdpModeToFlag(xdpMode)); err != nil {
+		// Program replacement unsuccessful, revert bpffs migration.
+		if err := bpf.FinalizeBPFFSMigration(bpf.MapPrefixPath(), spec, true); err != nil {
```
Looks like `err` is still being shadowed here, is that intentional?

https://github.com/cilium/cilium/pull/19159/files#r887403069

ti-mo replied:
Ah, thought you wanted to leave this as-is. We'll log the migration error and return the netlink one. The failed migration isn't actually the error we're trying to recover from.

(also, nit: the shadowing is fine, I assume the real problem here is swallowing the attach error)

joestringer replied:
Ah yes this is fine sorry. I couldn't quite tell what the resolution was from the previous time.

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -294,23 +294,23 @@ func (e *Endpoint) updateAndOverrideEndpointOptions(opts option.OptionMap) (opts
 
 // Called with e.mutex UNlocked
-func (e *Endpoint) regenerate(context *regenerationContext) (retErr error) {
+func (e *Endpoint) regenerate(ctx *regenerationContext) (retErr error) {
```
Please consider putting purely cosmetic changes like this in their own commits, just to simplify review when there are some changes that are functional vs. others that are not functional. I agree that this is a nice cleanup though, `ctx` should be the standard var name for the context, mostly just to avoid the package name conflict in case that comes up.

ti-mo replied:
This was done out of necessity to be able to reference `context.Canceled` in `updateRegenerationStatistics()` and `regenerate()`, but I'll split this off into a separate commit.

joestringer reviewed (changes requested):
The last commit seems to be rewriting `FunctionQueue`, which I don't think necessarily needs to happen to this degree but I guess it's fine. That said, there's a few things that could be improved with the new API, which should be fixed up if you are planning to rewrite that structure. Minor stuff, comments below.

I noticed a couple of odd things about the logging usage and errors which I think is important to address just to ensure that we don't start hiding real errors in production environments. I tried to highlight them to the best of my ability, hopefully it's not too much effort to try out those changes and see whether they break any expectations in the CI.

I didn't look too closely at the earlier commits since I had previously seen them and they seemed fine at that point.

ti-mo committed:
datapath: remove unused serializer retry logic

This commit removes the serializer's retry behavior since it was left unused.
The only queue that was ever created only had a single slot, and the only
user of the queue (fetchOrCompile) would unconditionally stop the queue.

In other words: a 1-slot queue that will only ever be used to process a single
event. This aims to simplify the implementation so it can eventually be
replaced by sync.Once or singleflight, which serve a similar purpose.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
endpoint: rename context parameter to ctx in Endpoint methods

ctx is the canonical name of a context given to a Go function.
Change this to make follow-up commits shorter.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
datapath: return callee's errors from fc.Wait()

When multiple Endpoints with the same template config are being (re)generated,
the 'active' one (with an in-flight clang invocation) might receive a context
cancellation.

When this happens, all other 'waiting' Endpoints would see fq.stopCh being
closed, and since they all have different (non-cancelled) Contexts, they
would all return from fq.Wait() without error. This results in many
'Could not locate previously compiled BPF template' errors in CI.

This patch changes error handling by returning the error from the enqueued
function instead of checking ctx.Err() explicitly in fq.Wait(). Wait() no
longer takes a context since selecting on its done channel races with the
write to fq.err.

Signed-off-by: Timo Beckers <timo@isovalent.com>

joestringer commented on pkg/serializer/func_queue.go:
```diff
@@ -3,100 +3,50 @@
 }
 
 // NewFunctionQueue returns a FunctionQueue that will be used to execute
 // functions in the same order they are enqueued.
-func NewFunctionQueue(queueSize uint) *FunctionQueue {
+func NewFunctionQueue() *FunctionQueue {
```
nit: The comment here implies that multiple functions will be executed ("same order they are enqueued") but at this point only one function gets executed.

Arguably it's not a `FunctionQueue` any more since it just runs the first function that's sent to it. (Is a queue a queue if the size is 1? Maybe, but the key property is not really useful)

These are cosmetic changes though, so feel free to either fix this up in an extra patch here or follow up separately.

ti-mo replied:
This `FunctionQueue` will be removed in subsequent work (see `replace with sync.Once` comment). Will scope this out in a few GH issues, as there's a lot to cover.

Marking as ready to merge.
